Feature/multiplex

iijung opened the pull request:
### 관련 이슈 
- #8 

### 설명
- 자세한 사용법은 `test/multiplex`에 추가한 server, client를 확인해주세요!
```sh
$ cd test/multiplex
$ make
$ ./main
$ make test   # nc -zv localhost 8080 으로 연결 확인
$ make test2  # netcat localhost 8080 으로 송수신 확인
```

### 참고
- 23.11.07 리눅스 환경에서만 테스트 완료

### Github 관련 참고
아직 맥 환경에서 테스트가 필요하므로 draft(초안)로 풀리퀘스트 올렸습니다.
- `This pull request is still a work in progress`: 이 풀 요청은 아직 진행 중인 작업입니다.
- `Draft pull requests cannot be merged.`: 초안 풀 요청은 병합 불가능합니다.
- `Ready for Review` 버튼을 눌러야 일반 풀 요청으로 변경됩니다. 

테스트 끝나고 `Ready for Review` 누를 예정인데 그 전에 리뷰 미리 남겨주셔도 됩니다!

iijung committed:
feat: IO multiplex (kqueue, epoll)

iijung committed:
test: add multiplexing test

iijung committed:
fix: kevent don't use bit flags

iijung committed:
fix: netcat command not found error

iijung committed:
fix: undeclare macro in bsd

iijung commented:
Give me feedback!

juwkim commented on libs/multiplex/io_event.hpp:
```diff
@@ -0,0 +1,30 @@
+#pragma once
+#ifndef LIBS_MULTIPLEX_IO_EVENT_HPP_
+# define LIBS_MULTIPLEX_IO_EVENT_HPP_
```
https://learn.microsoft.com/ko-kr/cpp/preprocessor/once?view=msvc-170
다중 포함 최적화를 2가지 중복하여 사용하는 이유가 있나요?
지금까지 한 리뷰와 동일한 관점에서, pragame once를 해석하지 못하는 컴파일러가 존재할 수 있습니다.
그리고 부연설명으로, 성능면에서 pragma once 구문이 전처리기 구문 (ifndef ~ define)보다 성능이 뛰어납니다.

iijung replied:
`#pragma once`를 지원하지 않는 경우 경고만 발생시키며, include guard를 통해 중복 헤더를 방지할 수 있습니다.

그렇지만 저희가 지키기로 한 [`c++ google style guide`](https://google.github.io/styleguide/cppguide.html)에서 `#pragma once`를 권장하지 않으니 제거 하는게 좋을 것 같네요.

juwkim commented on libs/multiplex/multiplex.cpp:
```diff
@@ -0,0 +1,121 @@
+# include <sys/epoll.h>
+#else
+# include <sys/event.h>
+#endif
```
전처리기 endif 구문 뒤에, // __LINUX__를 붙여줘야 가독성이 올라갑니다.
지금과 같은 짧은 코드에서는 관계가 없지만, 긴 코드에서는 어느 ifdef에 매칭되는 endif문 인지 찾기 어려울 수 있어요.
관용적으로 붙이고 있습니다.

iijung replied:
헤더 파일에는 꼬리 주석을 남기고 있지만 이 부분은 코드가 복잡하지 않아서 필요성을 느끼지 못 했습니다.

수정이 필요한 경우 꼬리 주석을 남기기보다는 multiplex_linux.cpp multiplex_macos.cpp로 파일을 분리한 뒤 Makefile에서 어떤 소스코드를 사용할지 선택하는 방식이 더 좋을 것 같습니다. 

개인적으로는 코드 유사성이 많고, 줄 수가 적으므로 지금 상태도 나쁘지 않다고 생각합니다.

juwkim replied:
네~ 코드 가독성에 따라 추가하는 방식 좋습니다!

juwkim commented on libs/multiplex/multiplex.cpp:
```diff
@@ -0,0 +1,121 @@
+	handler_ = epoll_create1(0);
+#else
+	handler_ = kqueue();
+#endif
```
이부분도 같은 내용입니다.

iijung replied:
상동

juwkim replied:
넵

juwkim commented on libs/multiplex/multiplex.cpp:
```diff
@@ -0,0 +1,121 @@
+
+void Multiplex::Loop(void)
+{
+	static const int	kMaxEvent = 10;
```
함수 내부에서 정적 상수를 선언하면, 메모리를 어딘가에서 잡고 있겠죠?
단순히 class내부에서 상수를 이용할 때, header에 선언한 방법도 있지만 동일한 문제로 사용되고 있지 않습니다.
약간의 트릭으로 enum을 이용해서 class 내부에서 상수를 메모리 낭비 없이 쓰는 방법이 있어요.

iijung replied:
이 부분은 `#define` vs `enum` vs `static const` 에 대해 고민을 많이 했는데 [스택오버플로우](https://stackoverflow.com/questions/1674032/static-const-vs-define-vs-enum)의 내용을 참고했습니다.
> If you were asking about C++, then you'd use option (1) — the static const — every time.

여러 논의가 오갔는데 enum을 선호하기는 하지만 안전성에 대한 언급이 계속 발생해서 결국 static const로 ...

juwkim replied:
그렇네요!!

juwkim commented on test/multiplex/Makefile:
```diff
@@ -0,0 +1,30 @@
+	main.cpp
+
+LIBS	= \
+	../../libs/multiplex/libmultiplex.a \
```
$(TOPDIR)/libs/multiplex/libmultiplex.a 을 쓰는 것은 어떨까요?

iijung replied:
해당 요청과 관련 없이 전체적으로 적용되어야 하므로 별도의 이슈로 등록했습니다. 이 부분은 반영 후 전체적으로 수정하겠습니다.
- #12

juwkim commented on test/multiplex/client.cpp:
```diff
@@ -0,0 +1,135 @@
+{
+}
+
+Client::Client(int fd, ListenConfig *conf): conf_(conf)
```
const int fd

juwkim commented on test/multiplex/client.cpp:
```diff
@@ -0,0 +1,135 @@
+	if (identifier_ == -1)
+		return ;
+	char buf[1025] = {0};
+	ssize_t nbytes = recv(identifier_, buf, sizeof(buf) - 1, 0);
```
const ssize_t

juwkim commented on test/multiplex/client.cpp:
```diff
@@ -0,0 +1,135 @@
+{
+	if (identifier_ == -1)
+		return ;
+	size_t len = sendbuf_.length();
```
const size_t

juwkim commented on test/multiplex/client.cpp:
```diff
@@ -0,0 +1,135 @@
+	size_t len = sendbuf_.length();
+	if (len == 0)
+		return ;
+	ssize_t nbytes = send(identifier_, sendbuf_.c_str(), len, 0);
```
const ssize_t

juwkim commented on test/multiplex/listen_config.hpp:
```diff
@@ -0,0 +1,29 @@
+#ifndef LISTEN_CONFIG_HPP_
+# define LISTEN_CONFIG_HPP_
+
+struct ListenConfig {
```
구조체가 불필요한 패딩으로 인해 크기가 증가했어요ㅠㅠ 패딩을 최대한 줄여주세요.

iijung replied:
multiplex 라이브러리 사용에 대해 간단하게 작성한 테스트 코드입니다.
현재 config에서 정보를 가져올 수 없어 만든 임시 구조체이므로 추후 사용하지 않을 예정이니 수정할 필요성을 못 느끼겠습니다.

juwkim commented on test/multiplex/server.cpp:
```diff
@@ -0,0 +1,165 @@
+
+int	Server::GetAddrInfo(struct addrinfo **info)
+{
+	struct addrinfo hints;
```
https://en.cppreference.com/w/c/language/struct_initialization
구조체를 생성과 동시에 초기화 할 수 있는 방법이 있어요.
class의 initializer list처럼 구조체도 존재합니다.

iijung replied:
해당 구조체는 C 라이브러리에서 제공하는 구조체인데 ={0} 사용 시 컴파일 오류가 발생하여 제거하고 각 요소를 참조할 겸 대해 직접 설정했습니다.

iijung replied:
- 136ad54e74ed40eddd3416116494e0b9e448f248 에 반영했습니다.

iijung replied:
반영한 거 revert 했습니다

juwkim commented on test/multiplex/server.cpp:
```diff
@@ -0,0 +1,165 @@
+
+	const char *node = conf_->address.c_str();
+	const char *service= conf_->port.c_str();
+	int ret = getaddrinfo(node, service, &hints, info);
```
const를 안 붙인 이유는 함수의 return 값이 int이기 때문이겠죠?

iijung replied:
아뇨 정확한 이유는 새벽에 거기까지 디테일을 챙기기에는 귀찮았기 때문입니다.
위 코드는 seojo님에게 참고자료를 만들어 주기 위해 새벽에 대충 만든 테스트 코드입니다.

server 내부에 있는 함수들은 원래  Open함수에 모두 포함이 되어 있었고 int ret은 아래와 같이 함수 결과를 저장하기 위한 일종의 임시 변수였습니다. 보기 쉽게 함수를 분리만 하고 해당 변수는 더 이상 신경 쓰지 않았습니다.
```
int ret;
ret = getaddrinfo ....
if (ret == -1) ... 
ret = this->CreatSocket();
if (ret != -1) .... d
```
함수가 분리된 시점에서 더 이상 변경되지 않으니 `const int` 붙이는 건 좋아보입니다.

iijung committed:
style: int -> const int

iijung committed:
chore: remove #pragma once in header

iijung committed:
chore: build error in libconfig.a

```sh
/Applications/Xcode.app/Contents/Developer/usr/bin/make libconfig.a
c++ -Wall -Wextra -Werror -Wpedantic -std=c++98 -MMD -MP  -I.    -c -o .objs/config.o config.cpp
In file included from config.cpp:1:
./config.hpp:37:24: error: no newline at end of file [-Werror,-Wnewline-eof]
                       ^
1 error generated.
make[3]: *** [.objs/config.o] Error 1
make[2]: *** [all] Error 2
```

iijung committed:
chore: specify TOPDIR in the first line

iijung committed:
refactor: init with structure declaration

iijung committed:
Revert "refactor: init with structure declaration"

This reverts commit 136ad54e74ed40eddd3416116494e0b9e448f248.

since c++20....

joseoungmin commented on libs/multiplex/io_event.cpp:
```diff
@@ -0,0 +1,53 @@
+	identifier_ = -1;
+}
+
+void IOEvent::Broken(int errcode)
+{
+	errno = errcode;
+	if (errno)
+		perror("Broken");
+	errno = 0;
+	this->Close();
+}
```
읽기 또는 쓰기 작업 후에 errno 값을 확인하는 것은 금지된다.
이 요구사항이 있는데 문제없는건가요?
perror 함수는 사용불가 함수 인 것 같아요

iijung replied:
허용 함수  확인을 못 했네요! 수정해놓겠습니다
streeror을 사용하면 errno를 사용할 필요는 없을 것 같아요. Broken(errcode)가 호출되는 곳을 보면 errno가 아니라 다른 value를 가져오고 있습니다

joseoungmin reviewed (commented):
깔끔하고 멋진 코드 잘 보았습니다 >.<

iijung commented:
fd가 -1인 이벤트 제거해야하는데 누락된 코드가 있어서 추가 필요!! 

- 4ec5a90536d334afd151832061e0f0a4e0b0a5c2 에서 반영 완료!

iijung committed:
fix: perror is prohibited.

iijung committed:
feat: if close socket, delete IOEvent in Multiplex

joseoungmin reviewed (commented):
흠잡을 곳 없는 깔끔한 코드인 것 같습니다
